repm/api_test.go

```diff
@@ -48,37 +50,75 @@ func TestBasics(t *testing.T) {
-		{"getRoot", `{}`, `{"root":"80744gunjjnad4e5pb8gkf8ntjcj02ca"}`, ""},                        // getRoot when db did change
+		{"put", `{"id": "foo", "value": "bar"}`, "", "Missing transaction ID"},
+
+		{"BeginTransaction", `{}`, `{"transactionId":1}`, ""},
+		{"put", `{"transactionId": 1, "id": "foo", "value": "bar"}`, `{}`, ""},
+		{"put", `{"transactionId": 1, "id": "foo"}`, ``, "value field is required"},
+		{"put", `{"transactionId": 1, "id": "foo", "value": null}`, `{}`, ""},
+		{"put", `{"transactionId": 1, "id": "foo", "value": "bar"}`, `{}`, ""}, // so we can scan it
+		{"getRoot", `{}`, `{"root":"4p3l8m7gjkkd8g3g0glothm038s61123"}`, ""},   // getRoot when db did change
+		{"commitTransaction", `{"transactionId":1}`, `{"ref":"p34f8g8jghkainifnsp966oqgf3pv88t"}`, ""},
+		{"getRoot", `{}`, `{"root":"d5024qks1v8sk57tjfg7ml14nugdm8e1"}`, ""}, // getRoot when db did change
 
 		// has
+
 		{"has", invalidRequest, ``, invalidRequestError},
-		{"has", `{"id": "foo"}`, `{"has":true}`, ""},
+		{"has", `{"id": "foo"}`, ``, "Missing transaction ID"},
+		{"BeginTransaction", `{}`, `{"transactionId":2}`, ""},
+		{"has", `{"transactionId": 2, "id": "foo"}`, `{"has":true}`, ""},
+		{"closeTransaction", `{"transactionId": 2}`, `{}`, ""},
 
 		// get
 		{"get", invalidRequest, ``, invalidRequestError},
-		{"get", `{"id": "foo"}`, `{"has":true,"value":"bar"}`, ""},
+		{"get", `{"id": "foo"}`, "", "Missing transaction ID"},
+		{"BeginTransaction", `{}`, `{"transactionId":3}`, ""},
+		{"get", `{"transactionId": 3, "id": "foo"}`, `{"has":true,"value":"bar"}`, ""},
+		{"closeTransaction", `{"transactionId": 3}`, `{}`, ""},
 
 		// scan
-		{"put", `{"id": "foopa", "value": "doopa"}`, `{"root":"i3p2c676665as6vhcv5032bhtguci02s"}`, ""},
-		{"scan", `{"prefix": "foo"}`, `[{"id":"foo","value":"bar"},{"id":"foopa","value":"doopa"}]`, ""},
-		{"scan", `{"start": {"id": {"value": "foo"}}}`, `[{"id":"foo","value":"bar"},{"id":"foopa","value":"doopa"}]`, ""},
-		{"scan", `{"start": {"id": {"value": "foo", "exclusive": true}}}`, `[{"id":"foopa","value":"doopa"}]`, ""},
+		{"BeginTransaction", `{}`, `{"transactionId":4}`, ""},
+		{"put", `{"transactionId": 4, "id": "foopa", "value": "doopa"}`, `{}`, ""},
+		{"commitTransaction", `{"transactionId":4}`, `{"ref":"n5sg13l13g5odv9kla1r5r5k7lhlef74"}`, ""},
+		{"BeginTransaction", `{}`, `{"transactionId":5}`, ""},
+		{"scan", `{"transactionId": 5, "prefix": "foo"}`, `[{"id":"foo","value":"bar"},{"id":"foopa","value":"doopa"}]`, ""},
+		{"scan", `{"transactionId": 5, "start": {"id": {"value": "foo"}}}`, `[{"id":"foo","value":"bar"},{"id":"foopa","value":"doopa"}]`, ""},
+		{"scan", `{"transactionId": 5, "start": {"id": {"value": "foo", "exclusive": true}}}`, `[{"id":"foopa","value":"doopa"}]`, ""},
+		{"closeTransaction", `{"transactionId":5}`, `{}`, ""},
 
 		// TODO: other scan operators
 	}
 
+	currentTXID := 0
+	re := regexp.MustCompile(`"transactionId":\s*(\d+)`)
+	findTransactionID := func(s string) {
```

Comment:
Isn't the transaction ID consistent across runs of the test? The state is per-connection so I don't see how it can change and why this string stuff is needed.

Reply:
It wasn't before. Let me strip this cruft